[Bug]: Encoding issue when reading BQ query result set in DataflowRunner

### What happened?

I tried to run a pipeline that ran a SQL query with a BQ DATE field in the result set with DataflowRunner and it resulted in an encoding error.  The exception message suggests that the correct coder was chosen (The Date logical type resolves to a Long) but the value resolution doesn't appear to happen with DataflowRunner.  It did with DirectRunner.


Affected Runners: DataflowRunner (DirectRunner with 2.43.0 and previous versions was fine)
Affected Versions: 2.43.0 (2.41.0, 2.42.0 do not appear to exhibit this behaviour.)

Sample Code

```
import com.google.api.services.bigquery.model.TableRow;
import org.apache.beam.runners.dataflow.options.DataflowPipelineOptions;
import org.apache.beam.sdk.Pipeline;
import org.apache.beam.sdk.io.gcp.bigquery.BigQueryIO;
import org.apache.beam.sdk.options.PipelineOptionsFactory;
import org.apache.beam.sdk.transforms.DoFn;
import org.apache.beam.sdk.transforms.MapElements;
import org.apache.beam.sdk.transforms.ParDo;
import org.apache.beam.sdk.values.PCollection;
import org.apache.beam.sdk.values.Row;
import org.apache.beam.sdk.values.TypeDescriptors;

import java.io.File;
import java.util.Arrays;
import java.util.stream.Collectors;

public class TestPipeline {
    public static void main(String[] args) {
        PipelineOptionsFactory.register(DataflowPipelineOptions.class);
        DataflowPipelineOptions options = PipelineOptionsFactory.fromArgs(args).as(DataflowPipelineOptions.class);
        options.setFilesToStage(
                Arrays.stream(System.getProperty("java.class.path").
                        split(File.pathSeparator)).
                        map(entry -> (new File(entry)).toString()).collect(Collectors.toList()));


        String testQuery = "select cast(\"2022-12-21\" as date) dt";
        Pipeline pipeline = Pipeline.create(options);
        PCollection<TableRow> tableRows = pipeline.apply(BigQueryIO.readTableRowsWithSchema()
                .fromQuery(testQuery)
                .usingStandardSql()
        );
        PCollection<Row> rows = tableRows.apply(MapElements.into(TypeDescriptors.rows())
                .via(tableRows.getToRowFunction()))
                .setRowSchema(tableRows.getSchema());
        rows.apply("println", ParDo.of(new DoFn<Row, Row>() {
            @ProcessElement
            public void processElement(@Element Row row) {
                System.out.println("println row" + row);
            }
        })).setRowSchema(tableRows.getSchema());
        pipeline.run();
    }
}
```

Exception
```
Error message from worker: java.lang.IllegalArgumentException: Unable to encode element 'GenericData{classInfo=[f], {dt=2022-12-21}}' with coder 'SchemaCoder<Schema: Fields:
Field{name=dt, description=, type=LOGICAL_TYPE, options={{}}}
Encoding positions:
{dt=0}
Options:{{}}UUID: 9622aec6-90b4-4021-b112-a7e4d2abecef  UUID: 9622aec6-90b4-4021-b112-a7e4d2abecef delegateCoder: org.apache.beam.sdk.coders.Coder$ByteBuddy$0cxD4zOH@3ef9d47a'.
	org.apache.beam.sdk.coders.Coder.getEncodedElementByteSize(Coder.java:300)
	org.apache.beam.sdk.coders.Coder.registerByteSizeObserver(Coder.java:291)
	org.apache.beam.sdk.util.WindowedValue$FullWindowedValueCoder.registerByteSizeObserver(WindowedValue.java:642)
	org.apache.beam.sdk.util.WindowedValue$FullWindowedValueCoder.registerByteSizeObserver(WindowedValue.java:558)
	org.apache.beam.runners.dataflow.worker.IntrinsicMapTaskExecutorFactory$ElementByteSizeObservableCoder.registerByteSizeObserver(IntrinsicMapTaskExecutorFactory.java:403)
	org.apache.beam.runners.dataflow.worker.util.common.worker.OutputObjectAndByteCounter.update(OutputObjectAndByteCounter.java:128)
	org.apache.beam.runners.dataflow.worker.DataflowOutputCounter.update(DataflowOutputCounter.java:67)
	org.apache.beam.runners.dataflow.worker.util.common.worker.OutputReceiver.process(OutputReceiver.java:43)
	org.apache.beam.runners.dataflow.worker.SimpleParDoFn$1.output(SimpleParDoFn.java:285)
	org.apache.beam.runners.dataflow.worker.repackaged.org.apache.beam.runners.core.SimpleDoFnRunner.outputWindowedValue(SimpleDoFnRunner.java:275)
	org.apache.beam.runners.dataflow.worker.repackaged.org.apache.beam.runners.core.SimpleDoFnRunner.access$900(SimpleDoFnRunner.java:85)
	org.apache.beam.runners.dataflow.worker.repackaged.org.apache.beam.runners.core.SimpleDoFnRunner$DoFnProcessContext.output(SimpleDoFnRunner.java:423)
	org.apache.beam.runners.dataflow.worker.repackaged.org.apache.beam.runners.core.SimpleDoFnRunner$DoFnProcessContext.output(SimpleDoFnRunner.java:411)
	org.apache.beam.sdk.io.gcp.bigquery.PassThroughThenCleanup$IdentityFn.processElement(PassThroughThenCleanup.java:84)
Caused by: java.lang.ClassCastException: java.time.LocalDate cannot be cast to java.lang.Long
	org.apache.beam.sdk.coders.VarLongCoder.encode(VarLongCoder.java:35)
	org.apache.beam.sdk.schemas.SchemaCoderHelpers$LogicalTypeCoder.encode(SchemaCoderHelpers.java:89)
	org.apache.beam.sdk.coders.RowCoderGenerator$EncodeInstruction.encodeDelegate(RowCoderGenerator.java:333)
	org.apache.beam.sdk.coders.Coder$ByteBuddy$0cxD4zOH.encode(Unknown Source)
	org.apache.beam.sdk.coders.Coder$ByteBuddy$0cxD4zOH.encode(Unknown Source)
	org.apache.beam.sdk.schemas.SchemaCoder.encode(SchemaCoder.java:124)
	org.apache.beam.sdk.coders.Coder.getEncodedElementByteSize(Coder.java:297)
	org.apache.beam.sdk.coders.Coder.registerByteSizeObserver(Coder.java:291)
	org.apache.beam.sdk.util.WindowedValue$FullWindowedValueCoder.registerByteSizeObserver(WindowedValue.java:642)
	org.apache.beam.sdk.util.WindowedValue$FullWindowedValueCoder.registerByteSizeObserver(WindowedValue.java:558)
	org.apache.beam.runners.dataflow.worker.IntrinsicMapTaskExecutorFactory$ElementByteSizeObservableCoder.registerByteSizeObserver(IntrinsicMapTaskExecutorFactory.java:403)
	org.apache.beam.runners.dataflow.worker.util.common.worker.OutputObjectAndByteCounter.update(OutputObjectAndByteCounter.java:128)
	org.apache.beam.runners.dataflow.worker.DataflowOutputCounter.update(DataflowOutputCounter.java:67)
	org.apache.beam.runners.dataflow.worker.util.common.worker.OutputReceiver.process(OutputReceiver.java:43)
	org.apache.beam.runners.dataflow.worker.SimpleParDoFn$1.output(SimpleParDoFn.java:285)
	org.apache.beam.runners.dataflow.worker.repackaged.org.apache.beam.runners.core.SimpleDoFnRunner.outputWindowedValue(SimpleDoFnRunner.java:275)
	org.apache.beam.runners.dataflow.worker.repackaged.org.apache.beam.runners.core.SimpleDoFnRunner.access$900(SimpleDoFnRunner.java:85)
	org.apache.beam.runners.dataflow.worker.repackaged.org.apache.beam.runners.core.SimpleDoFnRunner$DoFnProcessContext.output(SimpleDoFnRunner.java:423)
	org.apache.beam.runners.dataflow.worker.repackaged.org.apache.beam.runners.core.SimpleDoFnRunner$DoFnProcessContext.output(SimpleDoFnRunner.java:411)
	org.apache.beam.sdk.io.gcp.bigquery.PassThroughThenCleanup$IdentityFn.processElement(PassThroughThenCleanup.java:84)
	org.apache.beam.sdk.io.gcp.bigquery.PassThroughThenCleanup$IdentityFn$DoFnInvoker.invokeProcessElement(Unknown Source)
	org.apache.beam.runners.dataflow.worker.repackaged.org.apache.beam.runners.core.SimpleDoFnRunner.invokeProcessElement(SimpleDoFnRunner.java:211)
	org.apache.beam.runners.dataflow.worker.repackaged.org.apache.beam.runners.core.SimpleDoFnRunner.processElement(SimpleDoFnRunner.java:188)
	org.apache.beam.runners.dataflow.worker.SimpleParDoFn.processElement(SimpleParDoFn.java:340)
	org.apache.beam.runners.dataflow.worker.util.common.worker.ParDoOperation.process(ParDoOperation.java:44)
	org.apache.beam.runners.dataflow.worker.util.common.worker.OutputReceiver.process(OutputReceiver.java:49)
	org.apache.beam.runners.dataflow.worker.util.common.worker.ReadOperation.runReadLoop(ReadOperation.java:218)
	org.apache.beam.runners.dataflow.worker.util.common.worker.ReadOperation.start(ReadOperation.java:169)
	org.apache.beam.runners.dataflow.worker.util.common.worker.MapTaskExecutor.execute(MapTaskExecutor.java:83)
	org.apache.beam.runners.dataflow.worker.BatchDataflowWorker.executeWork(BatchDataflowWorker.java:420)
	org.apache.beam.runners.dataflow.worker.BatchDataflowWorker.doWork(BatchDataflowWorker.java:389)
	org.apache.beam.runners.dataflow.worker.BatchDataflowWorker.getAndPerformWork(BatchDataflowWorker.java:314)
	org.apache.beam.runners.dataflow.worker.DataflowBatchWorkerHarness$WorkerThread.doWork(DataflowBatchWorkerHarness.java:140)
	org.apache.beam.runners.dataflow.worker.DataflowBatchWorkerHarness$WorkerThread.call(DataflowBatchWorkerHarness.java:120)
	org.apache.beam.runners.dataflow.worker.DataflowBatchWorkerHarness$WorkerThread.call(DataflowBatchWorkerHarness.java:107)
	java.util.concurrent.FutureTask.run(FutureTask.java:266)
	org.apache.beam.sdk.util.UnboundedScheduledExecutorService$ScheduledFutureTask.run(UnboundedScheduledExecutorService.java:162)
	java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	java.lang.Thread.run(Thread.java:748)
```



### Issue Priority

Priority: 2 (default / most bugs should be filed as P2)

### Issue Components

- [ ] Component: Python SDK
- [X] Component: Java SDK
- [ ] Component: Go SDK
- [ ] Component: Typescript SDK
- [ ] Component: IO connector
- [ ] Component: Beam examples
- [ ] Component: Beam playground
- [ ] Component: Beam katas
- [ ] Component: Website
- [ ] Component: Spark Runner
- [ ] Component: Flink Runner
- [ ] Component: Samza Runner
- [ ] Component: Twister2 Runner
- [ ] Component: Hazelcast Jet Runner
- [ ] Component: Google Cloud Dataflow Runner

--- a/sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/SchemaTranslation.java
+++ b/sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/SchemaTranslation.java
@@ -439,9 +439,8 @@ public class SchemaTranslation {
                         logicalType.getPayload().toByteArray(), "logicalType"));
           } catch (IllegalArgumentException e) {
             LOG.warn(
-                String.format(
-                    "Unable to deserialize the logical type %s from proto. Mark as UnknownLogicalType",
-                    urn));
+                "Unable to deserialize the logical type {} from proto. Mark as UnknownLogicalType.",
+                urn);
           }
         }
         // assemble an UnknownLogicalType

--- a/sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/SchemaTranslation.java
+++ b/sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/SchemaTranslation.java
@@ -56,6 +56,7 @@ import org.apache.beam.sdk.schemas.logicaltypes.VariableString;
 import org.apache.beam.sdk.util.SerializableUtils;
 import org.apache.beam.sdk.values.Row;
 import org.apache.beam.vendor.grpc.v1p48p1.com.google.protobuf.ByteString;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.annotations.VisibleForTesting;
 import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.base.Preconditions;
 import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.collect.ImmutableMap;
 import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.collect.Iterables;
@@ -63,6 +64,8 @@ import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.collect.Maps;
 import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.io.ByteStreams;
 import org.apache.commons.lang3.ClassUtils;
 import org.checkerframework.checker.nullness.qual.Nullable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 /** Utility methods for translating schemas. */
 @Experimental(Kind.SCHEMAS)
@@ -71,6 +74,7 @@ import org.checkerframework.checker.nullness.qual.Nullable;
   "rawtypes"
 })
 public class SchemaTranslation {
+  private static final Logger LOG = LoggerFactory.getLogger(SchemaTranslation.class);
 
   private static final String URN_BEAM_LOGICAL_DECIMAL = FixedPrecisionNumeric.BASE_IDENTIFIER;
   private static final String URN_BEAM_LOGICAL_JAVASDK = "beam:logical_type:javasdk:v1";
@@ -124,8 +128,8 @@ public class SchemaTranslation {
         .build();
   }
 
-  private static SchemaApi.FieldType fieldTypeToProto(
-      FieldType fieldType, boolean serializeLogicalType) {
+  @VisibleForTesting
+  static SchemaApi.FieldType fieldTypeToProto(FieldType fieldType, boolean serializeLogicalType) {
     SchemaApi.FieldType.Builder builder = SchemaApi.FieldType.newBuilder();
     switch (fieldType.getTypeName()) {
       case ROW:
@@ -297,7 +301,8 @@ public class SchemaTranslation {
         .withDescription(protoField.getDescription());
   }
 
-  private static FieldType fieldTypeFromProto(SchemaApi.FieldType protoFieldType) {
+  @VisibleForTesting
+  static FieldType fieldTypeFromProto(SchemaApi.FieldType protoFieldType) {
     FieldType fieldType = fieldTypeFromProtoWithoutNullable(protoFieldType);
 
     if (protoFieldType.getNullable()) {
@@ -426,26 +431,33 @@ public class SchemaTranslation {
           return FieldType.DATETIME;
         } else if (urn.equals(URN_BEAM_LOGICAL_DECIMAL)) {
           return FieldType.DECIMAL;
-        } else if (urn.equals(URN_BEAM_LOGICAL_JAVASDK)) {
-          return FieldType.logicalType(
-              (LogicalType)
-                  SerializableUtils.deserializeFromByteArray(
-                      logicalType.getPayload().toByteArray(), "logicalType"));
-        } else {
-          @Nullable FieldType argumentType = null;
-          @Nullable Object argumentValue = null;
-          if (logicalType.hasArgumentType()) {
-            argumentType = fieldTypeFromProto(logicalType.getArgumentType());
-            argumentValue = fieldValueFromProto(argumentType, logicalType.getArgument());
+        } else if (urn.startsWith("beam:logical_type:")) {
+          try {
+            return FieldType.logicalType(
+                (LogicalType)
+                    SerializableUtils.deserializeFromByteArray(
+                        logicalType.getPayload().toByteArray(), "logicalType"));
+          } catch (IllegalArgumentException e) {
+            LOG.warn(
+                String.format(
+                    "Unable to deserialize the logical type %s from proto. Mark as UnknownLogicalType",
+                    urn));
           }
-          return FieldType.logicalType(
-              new UnknownLogicalType(
-                  urn,
-                  logicalType.getPayload().toByteArray(),
-                  argumentType,
-                  argumentValue,
-                  fieldTypeFromProto(logicalType.getRepresentation())));
         }
+        // assemble an UnknownLogicalType
+        @Nullable FieldType argumentType = null;
+        @Nullable Object argumentValue = null;
+        if (logicalType.hasArgumentType()) {
+          argumentType = fieldTypeFromProto(logicalType.getArgumentType());
+          argumentValue = fieldValueFromProto(argumentType, logicalType.getArgument());
+        }
+        return FieldType.logicalType(
+            new UnknownLogicalType(
+                urn,
+                logicalType.getPayload().toByteArray(),
+                argumentType,
+                argumentValue,
+                fieldTypeFromProto(logicalType.getRepresentation())));
       default:
         throw new IllegalArgumentException(
             "Unexpected type_info: " + protoFieldType.getTypeInfoCase());

--- a/sdks/java/core/src/test/java/org/apache/beam/sdk/schemas/SchemaTranslationTest.java
+++ b/sdks/java/core/src/test/java/org/apache/beam/sdk/schemas/SchemaTranslationTest.java
@@ -40,9 +40,16 @@ import org.apache.beam.sdk.schemas.Schema.Field;
 import org.apache.beam.sdk.schemas.Schema.FieldType;
 import org.apache.beam.sdk.schemas.logicaltypes.DateTime;
 import org.apache.beam.sdk.schemas.logicaltypes.FixedBytes;
+import org.apache.beam.sdk.schemas.logicaltypes.FixedPrecisionNumeric;
+import org.apache.beam.sdk.schemas.logicaltypes.FixedString;
 import org.apache.beam.sdk.schemas.logicaltypes.MicrosInstant;
+import org.apache.beam.sdk.schemas.logicaltypes.NanosDuration;
+import org.apache.beam.sdk.schemas.logicaltypes.NanosInstant;
 import org.apache.beam.sdk.schemas.logicaltypes.PythonCallable;
 import org.apache.beam.sdk.schemas.logicaltypes.SchemaLogicalType;
+import org.apache.beam.sdk.schemas.logicaltypes.SqlTypes;
+import org.apache.beam.sdk.schemas.logicaltypes.VariableBytes;
+import org.apache.beam.sdk.schemas.logicaltypes.VariableString;
 import org.apache.beam.sdk.values.Row;
 import org.apache.beam.vendor.grpc.v1p48p1.com.google.protobuf.ByteString;
 import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.base.Charsets;
@@ -392,6 +399,41 @@ public class SchemaTranslationTest {
 
       assertThat(exception.getMessage(), containsString("field_no_typeInfo"));
       assertThat(exception.getCause().getMessage(), containsString("TYPEINFO_NOT_SET"));
+    }
+  }
+
+  @RunWith(JUnit4.class)
+  public static class LogicalTypesTest {
+    @Test
+    public void testPortableLogicalTypeSerializeDeserilizeCorrectly() {
+      List<Schema.FieldType> testCases =
+          ImmutableList.<Schema.FieldType>builder()
+              .add(FieldType.logicalType(SqlTypes.DATE))
+              .add(FieldType.logicalType(SqlTypes.TIME))
+              .add(FieldType.logicalType(SqlTypes.DATETIME))
+              .add(FieldType.logicalType(SqlTypes.TIMESTAMP))
+              .add(FieldType.logicalType(new NanosInstant()))
+              .add(FieldType.logicalType(new NanosDuration()))
+              .add(FieldType.logicalType(FixedBytes.of(10)))
+              .add(FieldType.logicalType(VariableBytes.of(10)))
+              .add(FieldType.logicalType(FixedString.of(10)))
+              .add(FieldType.logicalType(VariableString.of(10)))
+              .add(FieldType.logicalType(FixedPrecisionNumeric.of(10)))
+              .build();
+
+      for (Schema.FieldType fieldType : testCases) {
+        SchemaApi.FieldType proto = SchemaTranslation.fieldTypeToProto(fieldType, true);
+        Schema.FieldType translated = SchemaTranslation.fieldTypeFromProto(proto);
+
+        assertThat(
+            translated.getLogicalType().getClass(), equalTo(fieldType.getLogicalType().getClass()));
+        assertThat(
+            translated.getLogicalType().getArgumentType(),
+            equalTo(fieldType.getLogicalType().getArgumentType()));
+        assertThat(
+            translated.getLogicalType().getArgument(),
+            equalTo(fieldType.getLogicalType().getArgument()));
+      }
     }
   }
 

--- a/sdks/java/core/src/test/java/org/apache/beam/sdk/schemas/SchemaTranslationTest.java
+++ b/sdks/java/core/src/test/java/org/apache/beam/sdk/schemas/SchemaTranslationTest.java
@@ -402,38 +402,42 @@ public class SchemaTranslationTest {
     }
   }
 
-  @RunWith(JUnit4.class)
+  /** Test schema translation of logical types. */
+  @RunWith(Parameterized.class)
   public static class LogicalTypesTest {
+    @Parameters(name = "{index}: {0}")
+    public static Iterable<Schema.FieldType> data() {
+      return ImmutableList.<Schema.FieldType>builder()
+          .add(FieldType.logicalType(SqlTypes.DATE))
+          .add(FieldType.logicalType(SqlTypes.TIME))
+          .add(FieldType.logicalType(SqlTypes.DATETIME))
+          .add(FieldType.logicalType(SqlTypes.TIMESTAMP))
+          .add(FieldType.logicalType(new NanosInstant()))
+          .add(FieldType.logicalType(new NanosDuration()))
+          .add(FieldType.logicalType(FixedBytes.of(10)))
+          .add(FieldType.logicalType(VariableBytes.of(10)))
+          .add(FieldType.logicalType(FixedString.of(10)))
+          .add(FieldType.logicalType(VariableString.of(10)))
+          .add(FieldType.logicalType(FixedPrecisionNumeric.of(10)))
+          .build();
+    }
+
+    @Parameter(0)
+    public Schema.FieldType fieldType;
+
     @Test
     public void testPortableLogicalTypeSerializeDeserilizeCorrectly() {
-      List<Schema.FieldType> testCases =
-          ImmutableList.<Schema.FieldType>builder()
-              .add(FieldType.logicalType(SqlTypes.DATE))
-              .add(FieldType.logicalType(SqlTypes.TIME))
-              .add(FieldType.logicalType(SqlTypes.DATETIME))
-              .add(FieldType.logicalType(SqlTypes.TIMESTAMP))
-              .add(FieldType.logicalType(new NanosInstant()))
-              .add(FieldType.logicalType(new NanosDuration()))
-              .add(FieldType.logicalType(FixedBytes.of(10)))
-              .add(FieldType.logicalType(VariableBytes.of(10)))
-              .add(FieldType.logicalType(FixedString.of(10)))
-              .add(FieldType.logicalType(VariableString.of(10)))
-              .add(FieldType.logicalType(FixedPrecisionNumeric.of(10)))
-              .build();
+      SchemaApi.FieldType proto = SchemaTranslation.fieldTypeToProto(fieldType, true);
+      Schema.FieldType translated = SchemaTranslation.fieldTypeFromProto(proto);
 
-      for (Schema.FieldType fieldType : testCases) {
-        SchemaApi.FieldType proto = SchemaTranslation.fieldTypeToProto(fieldType, true);
-        Schema.FieldType translated = SchemaTranslation.fieldTypeFromProto(proto);
-
-        assertThat(
-            translated.getLogicalType().getClass(), equalTo(fieldType.getLogicalType().getClass()));
-        assertThat(
-            translated.getLogicalType().getArgumentType(),
-            equalTo(fieldType.getLogicalType().getArgumentType()));
-        assertThat(
-            translated.getLogicalType().getArgument(),
-            equalTo(fieldType.getLogicalType().getArgument()));
-      }
+      assertThat(
+          translated.getLogicalType().getClass(), equalTo(fieldType.getLogicalType().getClass()));
+      assertThat(
+          translated.getLogicalType().getArgumentType(),
+          equalTo(fieldType.getLogicalType().getArgumentType()));
+      assertThat(
+          translated.getLogicalType().getArgument(),
+          equalTo(fieldType.getLogicalType().getArgument()));
     }
   }
 

